### ISSUE TITLE
Numerous needless hashes are wrapping the snapshot string

### ISSUE THREAD

**Issue by nyurik** [mitsuhiko/insta#600]
### What happened?

If a string contains an character that requires escaping in a non-raw literal, the number of hashes is incorrect.  This is especially noticeable when validating markdown generation.

### Reproduction steps

```rust
// No newlines, a single `\` (escaped)
let value = "abc \\######### xyz";
assert_snapshot!(value, @r##########"abc \######### xyz"##########);

// Same number of hashes, only this time there is no escapable symbols
let value = "abc ######### xyz";
assert_snapshot!(value, @"abc ######### xyz");

// A single newline
let value = "abc \n######### xyz";
assert_snapshot!(value, @r##########"
abc 
######### xyz
"##########);

// Copy of the previous test - note that newline is inserted before and after, plus indentation. I assume this is by design?
let value = "
abc
######### xyz
";
assert_snapshot!(value, @r##########"

    abc
    ######### xyz
    
"##########);
```

### Insta Version

1.40.0

### rustc Version

1.81.0

### What did you expect?

There should be as many hash characters as really needed.

**Comment by max-sixty** [comment]
Re:

```
// Copy of the previous test - note that newline is inserted before and after, plus indentation. I assume this is by design?
```

...this should be solved by https://github.com/mitsuhiko/insta/pull/581 (and tested / supplemented by https://github.com/mitsuhiko/insta/pull/563). These didn't get a review before the most recent release.

---

Re the hashes: we previously always used `###`, this was recently changed to base off the number of hashes in the string. There was a preference to always use raw strings (https://github.com/mitsuhiko/insta/pull/561#issuecomment-2283574586) — but we can rethink how we decide the number of hashes.

What rule would you propose?

**Comment by nyurik** [comment]
thx for the quick and thorough reply!

I think we should base it off of the definition of the raw string itself -- it ends when it sees a `"####` - so we just need to find the longest double quote+hashes string, and add one more hash to use as the string declaration.

regex dependency is currently optional, but could be made required, in which case something like this:

```rust
    let text = "foo ### bar";
    let re = Regex::new("\"#*").unwrap();

    let number_of_hashes_to_use = re
        .find_iter(text)
        .map(|m| m.as_str().len())
        .max()
        .unwrap_or(0);
```

**Comment by max-sixty** [comment]
Yes I think that could work quite well! 

We are really vigilant about dependencies but could do something similar without regex.

No pressure but if you wanted to do a PR that would be great; otherwise I'll try and get to it soom.

**Comment by max-sixty** [comment]
Fixed by #603
